src/audio/tdfb/tdfb.c

```diff
@@ -57,10 +63,58 @@ DECLARE_SOF_RT_UUID("tdfb", tdfb_uuid,  0xdd511749, 0xd9fa, 0x455c, 0xb3, 0xa7,
 
 DECLARE_TR_CTX(tdfb_tr, SOF_UUID(tdfb_uuid), LOG_LEVEL_INFO);
 
-/* IPC */
+/* IPC helpers for control update to user space */
 
-/* TODO: ALSA control update to user space need to be moved to module adapter */
+#if CONFIG_IPC_MAJOR_4
+static struct ipc_msg *tdfb_notification_init(struct processing_module *mod,
+					      uint32_t control_type_param_id,
+					      uint32_t control_id)
+{
+	struct ipc_msg msg_proto;
+	struct comp_dev *dev = mod->dev;
+	struct comp_ipc_config *ipc_config = &dev->ipc_config;
+	union ipc4_notification_header *primary =
+		(union ipc4_notification_header *)&msg_proto.header;
+	struct ipc_msg *msg;
+	struct tdfb_notification_payload *payload;
+
+	memset_s(&msg_proto, sizeof(msg_proto), 0, sizeof(msg_proto));
+	primary->r.notif_type = SOF_IPC4_MODULE_NOTIFICATION;
+	primary->r.type = SOF_IPC4_GLB_NOTIFICATION;
+	primary->r.rsp = SOF_IPC4_MESSAGE_DIR_MSG_REQUEST;
+	primary->r.msg_tgt = SOF_IPC4_MESSAGE_TARGET_FW_GEN_MSG;
+	msg = ipc_msg_w_ext_init(msg_proto.header, msg_proto.extension,
+				 sizeof(struct tdfb_notification_payload));
+	if (!msg)
+		return NULL;
+
+	payload = (struct tdfb_notification_payload *)msg->tx_data;
+	payload->module_data.instance_id = IPC4_INST_ID(ipc_config->id);
+	payload->module_data.module_id = IPC4_MOD_ID(ipc_config->id);
+	payload->module_data.event_id = SOF_IPC4_NOTIFY_MODULE_EVENTID_ALSA_MAGIC_VAL |
+		control_type_param_id;
+	payload->module_data.event_data_size = sizeof(struct sof_ipc4_control_msg_payload) +
```

Comment:
where do we validate size ?

Reply:
Sorry, what do you mean? The above  = ipc_msg_w_ext_init() has accepted the entire message size at this point of code. The used notification payload is fixed size, there is only one num_elems in used enum control. The kernel would error if the channel would not exist in topology.

Reply:
The FW should not trust the sender, i.e. it has to check the data is a valid size, so ```uint32_t event_data_size;``` has to be checked by FW.

Reply:
FW is in this case the sender and creates in this function a message template that is used for entire life time of it. When enum control is updated, only  the `ipc_payload->control_value.value` is updated in the next function tdfb_send_notification(). I think I'm missing something but I don't know how to make this more secure than this.

Reply:
This sounds fine is FW locally creates this data, the point is that remotely generated data is untrsuted and should always be validated.